lib/rpush/daemon/fcm/delivery.rb

```diff
@@ -0,0 +1,169 @@
+          fail Rpush::DeliveryError.new(403, @notification.id, 'FCM: The sender ID was mismatched. It seems the device token is wrong.')
+        end
+
+        def unregistered(response)
+          error = parse_error(response)
+          reflect(:fcm_invalid_device_token, @app, error, @notification.device_token)
+          fail Rpush::DeliveryError.new(404, @notification.id, "FCM: Client was not registered for your app. (#{error})")
+        end
+
+        def too_many_requests
+          fail Rpush::DeliveryError.new(429, @notification.id, 'FCM: Slow down. Too many requests were sent!')
+        end
+
+        def internal_server_error(response)
+          retry_delivery(@notification, response)
+          log_warn("FCM: responded with an Internal Error. " + retry_message)
+        end
+
+        def bad_gateway(response)
+          retry_delivery(@notification, response)
+          log_warn("FCM: responded with a Bad Gateway Error. " + retry_message)
+        end
+
+        def service_unavailable(response)
+          retry_delivery(@notification, response)
+          log_warn("FCM: responded with an Service Unavailable Error. " + retry_message)
+        end
+
+        def other_5xx_error(response)
+          retry_delivery(@notification, response)
+          log_warn("FCM: responded with a 5xx Error. " + retry_message)
+        end
+
+        def parse_error(response)
+          error = multi_json_load(response.body)['error']
+          puts "PARSED: #{error} #{error['message']}"
+          "#{error['status']}: #{error['message']}"
+        end
+
+        def deliver_after_header(response)
+          Rpush::Daemon::RetryHeaderParser.parse(response.header['retry-after'])
+        end
+
+        def retry_delivery(notification, response)
+          time = deliver_after_header(response)
+          if time
+            mark_retryable(notification, time)
+          else
+            mark_retryable_exponential(notification)
+          end
+        end
+
+        def retry_message
+          "Notification #{@notification.id} will be retried after #{@notification.deliver_after.strftime('%Y-%m-%d %H:%M:%S')} (retry #{@notification.retries})."
+        end
+
+        def obtain_access_token
+          if fetch_access_token?
+            token = fetch_access_token
+            @app.update(fcm_access_token: token, fcm_access_token_expiration: Time.zone.now + TOKEN_VALID_FOR_SEC)
+          end
+          @app.fcm_access_token
+        end
+
+        def fetch_access_token?
+          @app.fcm_access_token.nil? || @app.fcm_access_token_expiration.nil? || @app.fcm_access_token_expiration < Time.zone.now
+        end
+
+        def fetch_access_token
+          credentials = Google::Auth::ServiceAccountCredentials.make_creds(
+            json_key_io: StringIO.new(@json_private_key),
+            scope: AUTH_SCOPE
+          )
+
+          token_response = credentials.fetch_access_token!
+          token_response['access_token']
```

Comment:
[Question]: Should we do this fetch manually or should we include `googleauth` as a dev dependency?

@v7chan How were you imagining we should handle this? I think either option makes sense, but we already know `googleauth` is tested and works, so maybe having it as a dev dependency would be safer.

Reply:
I think having it as a runtime dependency is fine; Google recommends using their official libraries anyway (and I don't think we're really going to contribute this back upstream)

Reply:
Oops used the wrong term - Runtime dependency!

Reply:
@nat-aryucharoen Given this, we should probably add 'googleauth' as a `add_runtime_dependency` in `rpush.gemspec`